### PR TITLE
Feaure/in memory batching

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description
Notion: https://www.notion.so/lyrasis/Implement-proposed-solution-for-search_index_refresh-085989c0bd504f3f8e65eca8993066b7

* Remove unbounded counts on every job loop
* Remove the pressure of JSON serializing from the clster, and adds it to the application server

## Motivation and Context

The search_index_refresh job runs every 6 minutes on the Production systems.
This is a necessary script (maybe not with frequency) and puts a lot of CPU pressure on the RDS cluster

The refactoring here reduces the CPU usage of the script.
What could not be done is changing the `SELECT works JOIN workcoveragerecords` into a `SELECT workcoveragerecords JOIN works` query
This is because the job depends on the fact that there may be works WITHOUT a corresponding workcoveragerecords entry, which would imply a work needing coverage
This can be changed but will need an architectural change of how workcoveragerecords are created

## How Has This Been Tested?

All unit tests are passing
New unit tests are written
Real world data was used to compare results manually

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/core/test_external_search.py [hunk 1]
```diff
@@ -4484,3 +4578,38 @@ def bulk(self, docs, **kwargs):
         assert work == record.obj
         assert True == record.transient
         assert "There was an error!" == record.exception
+
+
+class DBStatementCounter(object):
```
Comment (io7m) [inline tests/core/test_external_search.py:4583]: Apparently we don't need to subclass `object` anymore in Python 3.

```
class DBStatementCounter:
```

Comment (RishiDiwanTT) [reply tests/core/test_external_search.py:4583]: This is true, this is stackoverflow code for counting statements, let me rectify
https://stackoverflow.com/a/33691585

File: core/model/work.py [hunk 2]
```diff
@@ -1428,6 +1432,282 @@ def assign_appeals(self, character, language, setting, story, cutoff=0.20):
     # that Elasticsearch can parse as a date.
     ELASTICSEARCH_TIME_FORMAT = 'YYYY-MM-DD"T"HH24:MI:SS"."MS'
 
+    @classmethod
```
Comment (io7m) [inline core/model/work.py:1435]: Was this code translated from somewhere?

Comment (RishiDiwanTT) [reply core/model/work.py:1435]: Yessir, to_search_documents() was an in-SQL query that did A LOT within the SQL cluster (query + aggregate + jsonify)
We've moved this to an "in-app" version, once approved I will rename THIS method to to_search_documents()
Methid is below in the same file 
https://github.com/ThePalaceProject/circulation/blob/532b7c79c9f1698b2e002a4daa8603c292334517/core/model/work.py#L1431

Comment (io7m) [reply core/model/work.py:1435]: Got it, thanks! I wanted to give this a quick read through to see if there were any obvious errors in translation. I'm guessing probably not given that the tests pass.

Comment (io7m) [reply core/model/work.py:1435]: Looks correct. :+1:

**Review by io7m — APPROVED** [review]

**Comment by tdilauro** [comment]
@RishiDiwanTT CI not passing. Looks like you need to merge the `main` branch to pick up the container fix for `lcpencrypt`.

**Comment by tdilauro** [comment]
@RishiDiwanTT I merged in `main` to pick up changes and force CI. Would you take a quick look to make sure that everything seems okay? If so, go ahead and merge.

**Comment by RishiDiwanTT** [comment]
@tdilauro thats weird because I had rebased main and pushed yesterday, not sure what went wrong

**Pull request merged**